added template code walk-through (WIP) (#2075)
nf-core/website: src/content/docs/contributing/nf_core_basic_training.md

# Explore nf-core/tools

The nf-core/tools package is already installed in the gitpod environment. Now you can check out which pipelines, subworkflows and modules are available via tools. To see all available commands of nf-core tools, run the following:

```bash
nf-core --help
```

We will touch on most of the commands for developers later throughout this tutorial.

# Create a pipeline from template

To get started with your new pipeline, run the create command:

```bash
nf-core create
```

This should open a command prompt similar to this:

```

                                          ,--./,-.
          ___     __   __   __   ___     /,-._.--~\
    |\ | |__  __ /  ` /  \ |__) |__         }  {
    | \| |       \__, \__/ |  \ |___     \`-._,-`-,
                                          `._,._,'

    nf-core/tools version 2.10 - https://nf-co.re

? Workflow name demotest
? Description test pipeline for demo
? Author FBo
? Do you want to customize which parts of the template are used? No

> INFO     Creating new nf-core pipeline: 'nf-core/demotest'                                                                                     INFO     Initialising pipeline git repository

Author: copy paste error, I'll fix it

│   └── local
│       └── input_check.nf
├── tower.yml
└── workflows
    └── demotest.nf
```

These are the files in detail:

1. **main.nf**

   This file contains the main nextflow pipeline code. Mostly this file is not touched.

2. **workflows/demotest.nf**

   This file is where the pipeline is going to be assembled. It connects the different modules and subworkflows.

3. **CHANGELOG.md, CODE_OF_CONDUCT.md, LICENSE, README.md, CITATIONS.md**

   These are standard files created for github repositories. As a default, this pipeline will be under an MIT licence. The CODE_OF_CONDUCT is specific for the nf-core community.

4. **assets/**

   This directory contains different templates such as email templates or the MultiQC config. In this course contents of this directory can be largely ignored.

5. **bin/**

   The `bin` directory contains executable scripts which are automatically added to the path by nextflow. As such, they can be called by name without using their absolut or relativ path. The python script `check_samplesheet.py` is part of the nf-core template, since typically, nf-core pipelines require a samplesheet as one of their inputs.

6. **conf/** and **nextflow.config**

>    The `nextflow.config` file is the main config file, it includes all the configurations in `conf/`. Importantly, `conf/` contains the `test.config` file, which is used for pipeline testing. In this course we are not going to touch config files, but they have been extensively covered in the following bytesize talks: [How nf-core configs work (nf-core/bytesize #2)](https://www.youtube.com/watch?v=cXBYusdjrc0&list=PL3xpfTVZLcNiF4hgkW0yXeNzr0d35qlIB&index=8&pp=gAQBiAQB), [Making a new institutional config profile (nf-core/bytesize #10)](https://www.youtube.com/watch?v=Ym1s6sKGzkw&list=PL3xpfTVZLcNiF4hgkW0yXeNzr0d35qlIB&index=9&pp=gAQBiAQB), [nf-core/bytesize: Using nf-core configs in custom pipelines](https://www.youtube.com/watch?v=zgcrI_0SUgg&list=PL3xpfTVZLcNiF4hgkW0yXeNzr0d35qlIB&index=40&pp=gAQBiAQB)

Reviewer: ```suggestion
   The `nextflow.config` file is the main config file. In addition to supplying default parameters, it imports all the configurations in `conf/`. Importantly, `conf/` contains the `test.config` file, which is used for pipeline testing. In this course we are not going to touch config files, but they have been extensively covered in the following bytesize talks: [How nf-core configs work (nf-core/bytesize #2)](https://www.youtube.com/watch?v=cXBYusdjrc0&list=PL3xpfTVZLcNiF4hgkW0yXeNzr0d35qlIB&index=8&pp=gAQBiAQB), [Making a new institutional config profile (nf-core/bytesize #10)](https://www.youtube.com/watch?v=Ym1s6sKGzkw&list=PL3xpfTVZLcNiF4hgkW0yXeNzr0d35qlIB&index=9&pp=gAQBiAQB), [nf-core/bytesize: Using nf-core configs in custom pipelines](https://www.youtube.com/watch?v=zgcrI_0SUgg&list=PL3xpfTVZLcNiF4hgkW0yXeNzr0d35qlIB&index=40&pp=gAQBiAQB)
```
Although includeConfig is the keyword in the config, I think imports is more understandable to the reader.

7. **docs/**

>    This directory contains additional information to the README file. The most important files are the `output.md` and the `usage.md` files. `usage.md` should descripe what exactly is needed to run the pipeline and `output.md` should descripe all outputs that can be expected. Importantly, for nf-core pipelines, the information from these two files will automatically be displayed on the nf-core website page of the pipeline.

Reviewer: ```suggestion
   This directory contains additional information to the README file. The most important files are the `output.md` and the `usage.md` files. `usage.md` should describe what exactly is needed to run the pipeline and `output.md` should describe all outputs that can be expected. Importantly, for nf-core pipelines, the information from these two files will automatically be displayed on the nf-core website page of the pipeline.
```

8. **lib/**

>    This directory contains groovy functions and classes that can be imported into the `main.nf` file.

Reviewer: ```suggestion
   This directory contains groovy functions and classes that are imported into the `main.nf` file to provide additional functionality not native to Nextflow.
```

9. **modules/local**

>    This is were all your custom non-nf-core modules go, we will see an example later in the course

Reviewer: ```suggestion
   This is where all your custom non-nf-core modules go. We will cover when and how to make local modules later in the course.
```

10. **modules/nf-core**

    All nf-core modules that are installed using the nf-core tooling will automatically show up in this directory. Keep them here, it is important for automatic updates.

11. **modules.json**

> 

Reviewer: ```suggestion

This file keeps track of modules installed using nf-core tools from the nf-core/modules repository. This file should only be updated using nf-core tools, and never manually.

```

12. **nextflow_schema.json**

    This file hosts all the parameters for the pipeline. Any new parameter should be added to this file using the `nf-core schema build` command. Similar to `output.md` and `usage.md`, the contents of `nextflow_schema.json` will get displayed on the pipeline page of nf-core pipelines.

13. **pyproject.toml**
14. **subworkflows/local**
15. **tower.yml**
16. **hidden directories and files**

    a) _.devcontainer/devcontainer.json_

    b) _.github/_

    Files in here are used for Continuous integration tests (CI) with github actions as well as other github related defaults, such as a template for issues. We will not touch on these in the course.

    c) _.gitignore_

    d) _.editorconfig_

    e) _.gitpod.yml_

> 

Reviewer: ```suggestion

This file provides settings to create a Cloud development environment in your browser using Gitpod. It comes installed with the tools necessary to develop and test nf-core pipelines, modules, and subworkflows, allowing you to develop from anywhere without installing anything locally.

```
